prefect-workflows.md

```diff
@@ -0,0 +1,103 @@
+
+- **Task scheduling:** Dask handles all task scheduling within a workflow, allowing Prefect to incentivize smaller tasks which Dask schedules with millisecond latency
+- **"Dataflow":** because Dask handles serializing and communicating the appropriate information between Tasks, Prefect can support "dataflow" as a first-class pattern
+- **Distributed computation:** Dask handles allocating Tasks to workers in a cluster, allowing users to immediately realize the benefits of distributed computation with minimal overhead
+- **Parallelism:** whether running in a cluster or locally, Dask provides parallel Task execution off the shelf
+
+Additionally, because Dask is written in pure Python and has an active open source community,
+we can very easily get feedback on possible bugs, and even contribute to improving the software ourselves.
+
+
+Pain points when using Dask
+---------------------------
+
+Our biggest pain point in using Dask has largely revolved around the ability (or lack
+thereof) to share futures between clients.  To provide a concrete example, suppose we start with a 
+list of numbers and, using [`client.map`](https://distributed.readthedocs.io/en/latest/api.html#distributed.Client.map)
+twice, we proceed to compute `x -> x + 1 -> x + 2` for each element of our list.  When using only dask primitives and a single client,
+these computations proceed asychronously, meaning that the final computation of each branch
+can begin without waiting on the other middle computations, as in this schematic:
+
+![Depth First Execution](depth-first.png)
+
+However, in Prefect, we aren't simply passing around Dask futures created from a single `Client` - when a [`map` operation](https://docs.prefect.io/guide/core_concepts/mapping.html#prefect-approach) occurs, the dask futures are actually created by a `worker_client` and attached to a Prefect `State` object.
+*Ideally*, we would leave these futures unresolved at this stage so that computation can proceed as above.  However, because 
+it is non-trivial to share futures between clients we must `gather` the futures with this same client, making
+our computation proceed in a "breadth-first" manner: 
+
+![Breadth first execution](breadth-first.png)
+
+This isn't the worst thing, but for longer pipelines it would be very nice to have the faster branches
+of the pipeline proceed with execution so that final results are produced earlier for inspection.
+
+Technology I use around Dask
+----------------------------
+
+Our preferred deployment of Prefect Flows uses [dask-kubernetes](https://github.com/dask/dask-kubernetes)
+to spin up a short-lived Dask Cluster in Kubernetes.  
+
+Otherwise, the logic contained within Prefect Tasks can be essentially arbitrary;
+many tasks in the system interact with databases, GCP resources, AWS, etc.
+
+
+Links
+-----
+
+- [Prefect Repo](https://github.com/PrefectHQ/prefect)
+- [Prefect on Dask Example](https://docs.prefect.io/guide/tutorials/dask-cluster.html)
```

Comment:
If interested in setting up a binder environment, we could also set up something on examples.dask.org ?

Reply:
Yea that's a great idea; I'll probably try to think of a more interesting example than just adding numbers (which is the "application" in that example), but I'll try to think of something fun / demo-worthy.